mathics/builtin/arithmetic.py

```diff
@@ -2205,3 +2212,131 @@ def apply(self, expr, evaluation):
             elif expr == SymbolFalse:
                 return Integer(0)
         return None
+
+
+class Assumptions(Predefined):
+    """
+    <dl>
+    <dt>'$Assumptions'
+      <dd>is the default setting for the Assumptions option used in such
+   functions as Simplify, Refine, and Integrate. 
+    </dl>
+    """
+
+    name = "$Assumptions"
+    attributes = ("Unprotected",)
+    rules = {
+        "$Assumptions": "True",
+    }
+
+
+class Assuming(Builtin):
+    """
+    <dl>
+    <dt>'Assuming[$cond$, $expr$]'
+      <dd>Evaluates $expr$ assuming the conditions $cond$
+    </dl>
+    >> $Assumptions = { x > 0 }
+     = {x > 0}
+    >> Assuming[y>0, $Assumptions]
+     = {x > 0, y > 0}
+    """
+
+    attributes = ("HoldRest",)
+
+    def apply_assuming(self, cond, expr, evaluation):
+        "Assuming[cond_, expr_]"
+        cond = cond.evaluate(evaluation)
+        if cond.is_true():
+            cond = []
+        elif cond.is_symbol() or not cond.has_form("List", None):
+            cond = [cond]
+        else:
+            cond = cond.leaves
+        assumptions = evaluation.definitions.get_definition(
+            "System`$Assumptions", only_if_exists=True
+        )
+
+        if assumptions:
+            assumptions = assumptions.ownvalues
+            if len(assumptions) > 0:
+                assumptions = assumptions[0].replace
+            else:
+                assumptions = None
+        if assumptions:
+            if assumptions.is_symbol() or not assumptions.has_form("List", None):
+                assumptions = [assumptions]
+            else:
+                assumptions = assumptions.leaves
+            cond = assumptions + tuple(cond)
+        Expression(
+            "Set", Symbol("System`$Assumptions"), Expression("List", *cond)
```

Comment:
Here is what I am seeing with this code:

```
In[1]:= $Assumptions
Out[1]= True

In[2]:= $Assumptions = x > 2
Out[2]= x > 2

In[3]:= $Assumptions
Out[3]= x > 2

In[4]:= Assuming[-10 < n < 10, Integrate[x^n, {x, 0, 1}]]
Out[4]= Piecewise[{{1 / (1 + n) - 0 ^ (1 + n) / (1 + n), n > -Infinity && n < Infinity && n != -1}, {Infinity, True}}]

In[5]:= $Assumptions
Out[5]= {x > 2}
```
Note that the type of `$Assumptions` changes, and the default `True` quickly gets changed to `{True}`.

But I think there is a deeper problem in that, if we ever support concurrency, then saving and restoring a variable will break things.

Is there is a way to wrap this in a context, so the scope of `$Assumptions` is narrowed to the expression getting evaluated and nothing needs to be saved and restored?

Reply:
One idea I had to handle this is to set a property in the  `Evaluation` object, or in `Definitions` object, each time $Assumptions is modified, or when Assuming is called. In any case, before looking at the best implementation, a way to evaluate predicates on expressions should be implemented.

Reply:
Another option to implement the temporal switching is to use `dynamic_scoping` (defined in mathics.builtin.scoping) as in `FindRoot`. This would make the code at least more readable.